jwt-post.md

```diff
@@ -0,0 +1,200 @@
+These tokens are generated by a cryptographically secure random-number generator.
+The front-end application reads the token in the header of each response and passes it on with the next request.
+For each request to a protected endpoint, the back end checks whether the two tokens match.
+If they match, the request is granted.
+Otherwise, it's denied.
+
+Keeping track of the CSRF tokens in the front end is not completely straightforward.
+It takes a little effort to keep track of the latest token value and forward it with each request, but that's an acceptable price to pay if you ask me.
+
+## Lifespan of a JWT
+
+Think about the following for a second: What happens to already issued JWTs when you change your credentials?
+What happens to already issued JWTs when you delete your account?
+In both scenarios, existing JWTs will remain valid.
+Without additional measures, JWTs remain valid until they expire or until the secret on the server is changed.
+If someone gets their hands on a token, it can be abused until it expires.
+If you want to invalidate a single token by changing the secret on the server, you invalidate all tokens.
+
+When should a JWT expire?
+On one hand, they should expire as soon as possible, to prevent misuse for long periods.
+On the other hand, they should expire as late as possible, so that users don't have to re-authenticate all the time.
+
+In practice, two types of tokens are used together, to achieve the best of both worlds.
+A short-lived *access* token is used for authentication per request.
+A long-lived *refresh* token is used to generate new access token when needed.
+
+Each time the refresh token is used to obtain a new access token, some additional checks could be made to enhance security.
+The refresh token can be used in combination with a blacklist, for example, to invalidate tokens that were issued for a particular user before a given point in time.
+
+## What kind of abuse is this protecting you from?
+
+Because the JWTs are stored in secure, HTTP-only cookies, it is implausible that someone would be able to access the JWTs themselves.
+An attacker would, for example, need access to a victim's computer to read the values of these cookies.
+The blacklist mentioned above could be used to invalidate JWTs comprised like this.
+However, if someone is able to access cookies directly from your computer, you have bigger problems to worry about that lie beyond the responsibility of an app developer.
+Moreover, there's no reasonable defense against someone willing to turn your life into a Quentin Tarantino movie to access your data or credentials.
+
+Other scenarios in which an attacker would be able to read the values of the JWTs would be when the attacker is able to intercept traffic between client and server or when an attacker would have access to the server.
+In such scenarios, all that can be done is patch up the security holes and change the secret key used to sign JWTs.
```

Comment:
> change the secret key used to sign JWTs

That's not really needed in this scenario, is it? The secret key itself is never exchanged between client and server.

Reply:
My reasoning is that you need to revoke all JWTs because you can't know which ones are compromised and need to assume that they're all compromised. Changing the secret key is the easiest way of revoking all the JWTs.

Do you agree? Or am I overlooking something?

Reply:
Ah yes, that makes sense.